packages/store/src/patterns/interface-tools.js

```diff
@@ -0,0 +1,227 @@
+import { Far } from '@endo/marshal';
+import { E } from '@endo/eventual-send';
+import { listDifference } from '@agoric/internal';
+
+import { fit } from './patternMatchers.js';
+
+const { details: X, quote: q } = assert;
+const { apply, ownKeys } = Reflect;
+const { defineProperties } = Object;
+
+const defendSyncArgs = (args, methodGuard, label) => {
+  const { argGuards, optionalArgGuards, restArgGuard } = methodGuard;
+  if (args.length < argGuards.length) {
+    assert.fail(
+      X`${label} args: ${args} - expected ${argGuards.length} arguments`,
+    );
+  }
+  for (let i = 0; i < args.length; i += 1) {
+    const arg = args[i];
+    const argLabel = `${label} arg ${i}`;
+    if (i < argGuards.length) {
+      fit(arg, argGuards[i], argLabel);
+    } else if (
+      optionalArgGuards &&
+      i < argGuards.length + optionalArgGuards.length
+    ) {
+      if (arg !== undefined) {
+        // In the optional section, an `undefined` arg succeeds
+        // unconditionally
+        fit(arg, optionalArgGuards[i - argGuards.length], argLabel);
+      }
+    } else if (restArgGuard) {
+      const restArg = harden(args.slice(i));
+      fit(restArg, restArgGuard, `${label} rest[${i}]`);
+      return;
+    } else {
+      assert.fail(
+        X`${argLabel}: ${args} - expected fewer than ${i + 1} arguments`,
+      );
+    }
+  }
+  if (restArgGuard) {
+    fit(harden([]), restArgGuard, `${label} rest[]`);
+  }
+};
+
+const defendSyncMethod = (method, methodGuard, label) => {
+  const { returnGuard } = methodGuard;
+  const { syncMethod } = {
+    // Note purposeful use of `this` and concise method syntax
+    syncMethod(...args) {
+      defendSyncArgs(harden(args), methodGuard, label);
+      const result = apply(method, this, args);
+      fit(harden(result), returnGuard, `${label}: result`);
+      return result;
+    },
+  };
```

Comment:
Should we be preserving names in `defendSyncMethod` and `defendAsyncMethod`?
```suggestion
  const { syncMethod } = {
    // Note purposeful use of `this` and concise method syntax
    syncMethod(...args) {
      defendSyncArgs(harden(args), methodGuard, label);
      const result = apply(method, this, args);
      fit(harden(result), returnGuard, `${label}: result`);
      return result;
    },
  };
  // Preserve the original method name.
  defineProperties(syncMethod, {
    name: { value: method.name, configurable: true },
  });
```

Reply:
Was already done at the bottom of `bindMethod`.